packages/fxa-settings/src/components/ThirdPartyAuth/index.tsx

```diff
@@ -147,48 +81,143 @@ const ThirdPartyAuth = ({ enabled = false }: ThirdPartyAuthProps) => {
+              responseType: 'code id_token',
+              accessType: 'offline',
+              prompt: 'consent',
+              responseMode: 'form_post',
+              onSubmit: onContinueWithApple,
+              buttonText: (
+                <>
+                  <AppleLogo />
+                  <FtlMsg id="continue-with-apple-button">
+                    Continue with Apple
+                  </FtlMsg>
+                </>
+              ),
             }}
-          >
-            <AppleLogo />
-            <FtlMsg id="continue-with-apple-button">Continue with Apple</FtlMsg>
-          </button>
+          />
         </div>
       </div>
     </>
   );
 };
 
+/**
+ * Represents the sign in form posted to google third party auth.
+ * Note that the innerRef is used by the parent component to trigger
+ * a form submission.
+ */
+const ThirdPartySignInForm = ({
+  party,
+  authorizationEndpoint,
+  state,
+  clientId,
+  scope,
+  redirectUri,
+  accessType,
+  prompt,
+  responseType,
+  responseMode,
+  buttonText,
+  onSubmit,
+}: {
+  party: 'google' | 'apple';
+  authorizationEndpoint: string;
+  state: string;
+  clientId: string;
+  scope: string;
+  redirectUri: string;
+  accessType: string;
+  prompt: string;
+  responseType: string;
+  responseMode?: string;
+  buttonText: ReactElement;
+  onSubmit?: FormEventHandler<HTMLFormElement>;
+}) => {
+  const stateRef = useRef<HTMLInputElement>(null);
+
+  function onClick() {
+    clearStoredParams();
+    stateRef.current!.value = getState();
+  }
+
+  if (onSubmit === undefined) {
+    onSubmit = () => true;
+  }
+
+  return (
+    <form action={authorizationEndpoint} method="GET" onSubmit={onSubmit}>
+      <input
+        data-testid={`${party}-signin-form-state`}
+        ref={stateRef}
+        type="hidden"
+        name="state"
+        value={state}
+      />
+      <input type="hidden" name="client_id" value={clientId} />
+      <input type="hidden" name="scope" value={scope} />
+      <input type="hidden" name="redirect_uri" value={redirectUri} />
+      <input type="hidden" name="access_type" value={accessType} />
+      <input type="hidden" name="prompt" value={prompt} />
+      <input type="hidden" name="response_type" value={responseType} />
+      {responseMode && (
+        <input type="hidden" name="response_mode" value={responseMode} />
+      )}
+
+      <button
+        type="submit"
+        className="w-full mt-2 justify-center text-black bg-transparent border-black border hover:border-grey-300 font-medium rounded-lg text-sm text-center inline-flex items-center"
+        onClick={onClick}
+      >
+        {buttonText}
+      </button>
+    </form>
+  );
+};
+
+function getState() {
+  // We stash originating location in the state oauth param
+  // because we will need it to use it to log the user into FxA
+  const currentParams = new URLSearchParams(window.location.search);
+  currentParams.delete('deeplink');
+  currentParams.set('showReactApp', 'true');
```

Comment:
@LZoog I think we should clean this up. We had a side bar on this previously. Just to rehash the convo here,  I think we shouldn't be using `URLSearchParams` directly. It's really designed to be wrapped with a model that has `@bind` and class-validator decorators. I feel like we should revisit this as a separate cleanup task since there are few places that I've spotted this so far.